### ISSUE TITLE
[BUG/Help] eval模式下，ChatGLM2模型对相同输入文本输出的logits数值不稳定（多次输入相同文本，输出logit不相同, ChatGLM1无此问题）

### ISSUE THREAD

**Issue by mzh1996** [THUDM/ChatGLM2-6B#302]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

with torch.no_grad():
    # 读取模型和tokenizer
    tokenizer = AutoTokenizer.from_pretrained(tokenizer_dir,trust_remote_code=True)
    model = AutoModel.from_pretrained(model_dir,trust_remote_code=True).cuda()
    model = model.eval()#这里已经是eval模式

    # 测试文本是以下text x 10
    text = "模型从第60+位置之后输出不稳定"
    input_str = ""
    for _ in range(10):
        input_str += text

    input_ids = tokenizer.encode(input_str)
    input_dict = {"input_ids":torch.tensor([input_ids]).cuda()}
    logger.info(str(input_dict["input_ids"].shape))

    logits_with_same_input = []

    # 相同的input_dict输入4次    
    dup_times = 4
    for _ in range(dup_times):
        res = model(**input_dict)
        logit = res["logits"]
        logger.info(str(logit.shape))
        # logit 的形状是 (1,seq_len,vocab_dim)
        # 我们先取logit每个位置的第0维进行观察
        logits_with_same_input.append(logit[0,:,0].cpu().numpy())

    # 看一下四次输出结果中 每个位置的logit数值是否完全一致
    # 如果不完全一致 打印出从第几个位置开始就不一致了
    logits_with_same_input_numpy_stack = np.stack(logits_with_same_input)
    logger.info(str(logits_with_same_input_numpy_stack.shape))
    # logits_with_same_input_numpy_stack.shape = (4, seq_len)

    for ind in range(logits_with_same_input_numpy_stack.shape[1]):
        tid = input_dict["input_ids"][0][ind]
        flag = ""
        set_diff = set(logits_with_same_input_numpy_stack[:,ind])
        if len(set_diff) != 1:
            flag = "diff!!!"
        logger.info(flag+str((ind,tokenizer.decode(tid))))

最后打印出的结果是从约第64个位置以后开始，输出就不一样了
更换输入text后结果也大致是如此（相同输入的logits每次不完全一样，也从第60+个位置开始就出现diff）

### Expected Behavior

在eval模式下，模型对相同输入的输出应该完全一致

### Steps To Reproduce

step1. 读取模型(6B,ChatGLM2,单卡,仿照cli_demo.py即可)，并开启eval模式(model.eval())
step2. 准备一条句子，长度在100字左右
step3. 构造句子的input_dict(包括input_ids等)，可以用官方的方法，也可以自己直接实现
step4. 将该input_dict输入到模型里 result = model(**input_dict)，取result中的logits
step5. 重复进行step4，收集每次的logits，然后逐个位置检查它们的数值是否一致
我这边发现不一致（更换了3条以上的句子测试，且都是从60+位置 大概是64/65位置开始出现不一致）

### Environment

```markdown
- OS:CentOS 7
- Python:3.10.4
- Transformers:4.30.2
- PyTorch:2.0.1
- CUDA Support (`python -c "import torch; print(torch.cuda.is_available())"`) : True
```

### Anything else?

我都是用的官方默认的config。因为出现不一致的位置（64）这个数值比较特殊，是我模型的配置没设置好吗？比如某个长度我设置成了64，导致模型即使是在eval模式下，对64以后的位置输出也可能不一致？

**Comment by mzh1996** [comment]
补充测试：

**Comment by lilongxian** [comment]
首先像这种每次推理得到的logist都会变，很大可能就在于dropout、L2等的使用上设置了固定值，没有区分训练和非训练的状态。
刚好，我在源码中找到了一处，如下：
class ChatGLMModel(ChatGLMPreTrainedModel):
。。。。。
        if self.pre_seq_len is not None:
            ............
            self.dropout = torch.nn.Dropout(0.1)

    def get_prompt(self, batch_size, device, dtype=torch.half):

       .........................
     
        past_key_values = self.dropout(past_key_values)
 
        return past_key_values

以上代码中的dropout被用在了前置编码器词嵌入方面：
past_key_values = self.embedding(prefix)
这是一个很重要的词嵌入，但是dropout比率被写死，训练中倒是应该维0.1，但在验证和预测推理应用中，无疑这个dropout=0.1会没运行一次模型裁剪你的前置编码词嵌入矩阵一次参数，你的logist必定会变！！！

所以：建议你有两种解决办法：
1. 等待官方确认这是不是bug，等官方修复后。
2. 你自己下载源码，自己修改一下，仅当训练时维0.1，非训练时为0.0即可。
3. 最简单的：放弃使用GLM2前置编码器即可，办法：将 --pre_seq_len参数改为None或者0

希望对你有帮助，欢迎大家踊跃讨论，一起为清华智谱GLM开源项目加油！

**Comment by mzh1996** [comment]
> 首先像这种每次推理得到的logist都会变，很大可能就在于dropout、L2等的使用上设置了固定值，没有区分训练和非训练的状态。 刚好，我在源码中找到了一处，如下： class ChatGLMModel(ChatGLMPreTrainedModel): 。。。。。 if self.pre_seq_len is not None: ............ self.dropout = torch.nn.Dropout(0.1)
> 
> ```
> def get_prompt(self, batch_size, device, dtype=torch.half):
> 
>    .........................
>  
>     past_key_values = self.dropout(past_key_values)
> 
>     return past_key_values
> ```
> 
> 以上代码中的dropout被用在了前置编码器词嵌入方面： past_key_values = self.embedding(prefix) 这是一个很重要的词嵌入，但是dropout比率被写死，训练中倒是应该维0.1，但在验证和预测推理应用中，无疑这个dropout=0.1会没运行一次模型裁剪你的前置编码词嵌入矩阵一次参数，你的logist必定会变！！！
> 
> 所以：建议你有两种解决办法：
> 
> 1. 等待官方确认这是不是bug，等官方修复后。
> 2. 你自己下载源码，自己修改一下，仅当训练时维0.1，非训练时为0.0即可。
> 3. 最简单的：放弃使用GLM2前置编码器即可，办法：将 --pre_seq_len参数改为None或者0
> 
> 希望对你有帮助，欢迎大家踊跃讨论，一起为清华智谱GLM开源项目加油！

非常感谢您的回复 我去检查一下，在不用ptuning的时候，会不会走到if self.pre_seq_len is not None: 这个分支里

**Comment by mzh1996** [comment]
> 首先像这种每次推理得到的logist都会变，很大可能就在于dropout、L2等的使用上设置了固定值，没有区分训练和非训练的状态。 刚好，我在源码中找到了一处，如下： class ChatGLMModel(ChatGLMPreTrainedModel): 。。。。。 if self.pre_seq_len is not None: ............ self.dropout = torch.nn.Dropout(0.1)
> 
> ```
> def get_prompt(self, batch_size, device, dtype=torch.half):
> 
>    .........................
>  
>     past_key_values = self.dropout(past_key_values)
> 
>     return past_key_values
> ```
> 
> 以上代码中的dropout被用在了前置编码器词嵌入方面： past_key_values = self.embedding(prefix) 这是一个很重要的词嵌入，但是dropout比率被写死，训练中倒是应该维0.1，但在验证和预测推理应用中，无疑这个dropout=0.1会没运行一次模型裁剪你的前置编码词嵌入矩阵一次参数，你的logist必定会变！！！
> 
> 所以：建议你有两种解决办法：
> 
> 1. 等待官方确认这是不是bug，等官方修复后。
> 2. 你自己下载源码，自己修改一下，仅当训练时维0.1，非训练时为0.0即可。
> 3. 最简单的：放弃使用GLM2前置编码器即可，办法：将 --pre_seq_len参数改为None或者0
> 
> 希望对你有帮助，欢迎大家踊跃讨论，一起为清华智谱GLM开源项目加油！

检查了一下，走不到那个分支（因为读取模型后默认pre_seq_len=0）。我手动将那个dropout的p调为0.0，发现也没解决。此外，我检查了config.json里，发现ChatGLM2已经默认把所有dropout的p置为0.0 。所以我个人认为排除了您说的2和3的可能。我比较认同您说的第一点

**Comment by mzh1996** [comment]
此外，这个问题有一个非常稳定的特性是，前64个字的logit是稳定一致的，从第65个以后开始，就不一样了。换了很多输入文本，都是这样，所以感觉是哪里有个类似64这种超参数，导致的这个问题？

**Comment by lilongxian** [comment]
@mzh1996  谢谢你关注！
这个问题我在看看，确实是，

**Comment by duzx16** [comment]
经过测试是 pytorch里的 `scaled_dot_product_attetion` 的加速实现带来的不确定性。很多 CUDA 实现会牺牲确定性来换取性能。如果一定要追求确定性的话可以通过
```python
torch.backends.cuda.enable_flash_sdp(False)
torch.backends.cuda.enable_mem_efficient_sdp(False)
```
来禁用加速实现，但是代价是显存占用会变高并且速度会变慢

**Comment by mzh1996** [comment]
> 经过测试是 pytorch里的 `scaled_dot_product_attetion` 的加速实现带来的不确定性。很多 CUDA 实现会牺牲确定性来换取性能。如果一定要追求确定性的话可以通过
> 
> ```python
> torch.backends.cuda.enable_flash_sdp(False)
> torch.backends.cuda.enable_mem_efficient_sdp(False)
> ```
> 
> 来禁用加速实现，但是代价是显存占用会变高并且速度会变慢

经过测试没有问题了！感谢各位专家的支持 @lilongxian @lilongxian

**Comment by jinzitian** [comment]
> > 经过测试是 pytorch里的 `scaled_dot_product_attetion` 的加速实现带来的不确定性。很多 CUDA 实现会牺牲确定性来换取性能。如果一定要追求确定性的话可以通过
> > ```python
> > torch.backends.cuda.enable_flash_sdp(False)
> > torch.backends.cuda.enable_mem_efficient_sdp(False)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 来禁用加速实现，但是代价是显存占用会变高并且速度会变慢
> 
> 经过测试没有问题了！感谢各位专家的支持 @lilongxian @duzx16

torch 1.12报错 module 'torch.backends.cuda' has no attribute 'enable_flash_sdp'，该怎么解决呢

**Comment by mzh1996** [comment]
> > > 经过测试是 pytorch里的 `scaled_dot_product_attetion` 的加速实现带来的不确定性。很多 CUDA 实现会牺牲确定性来换取性能。如果一定要追求确定性的话可以通过
> > > ```python
> > > torch.backends.cuda.enable_flash_sdp(False)
> > > torch.backends.cuda.enable_mem_efficient_sdp(False)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 来禁用加速实现，但是代价是显存占用会变高并且速度会变慢
> > 
> > 
> > 经过测试没有问题了！感谢各位专家的支持 @lilongxian @duzx16
> 
> torch 1.12报错 module 'torch.backends.cuda' has no attribute 'enable_flash_sdp'，该怎么解决呢

torch1.12应该不会出这个问题啊 因为torch1就没有flash attention
